certificates/src/lib.rs

```diff
@@ -108,6 +142,33 @@ impl CertificateValidator {
         // to use with verifying the private key.
         let cert = certs.swap_remove(0);
 
+        if let Some(hostname) = hostname {
+            let dns_names = cert_dns_names(&cert)?;
+            let hostname = DnsName::from_utf8(hostname)
+                .map_err(CertificateError::InvalidDnsName)?;
+            if !dns_names
+                .iter()
+                .any(|dns_name| dns_name_matches_hostname(dns_name, &hostname))
+            {
+                return Err(CertificateError::NoDnsNameMatchingHostname(
+                    hostname.to_string(),
+                ));
+            }
+        }
+
+        // We want to check whether this cert is suitable for use by a web
+        // server, which requires checking the x509 extended attributes.
+        // Unfortunately, the openssl crate does not expose these
+        // (https://github.com/sfackler/rust-openssl/issues/373), so we'll do
+        // something pretty gross: convert this cert to DER, then re-parse it
+        // via the pure Rust x509-parser crate, which does. We want to err on
+        // the side of "allow the cert" here; we're using a non-openssl crate
+        // and manually checking some SSL extensions, so we will only fail if we
+        // _know_ the cert is not valid.
+        if let Ok(cert_der) = cert.to_der() {
+            validate_cert_der_extended_key_usage(&cert_der)?;
+        }
```

Comment:
I am not sure that this check is worth it (but I don't feel strongly we should take it out, either).  I don't think I've ever run into such a certificate.

Reply:
I added this check specifically because @jmpesp generated a cert that fails this check but not the others. :sweat_smile: